### ISSUE TITLE
Allow passing extra arguments to key callbacks in the visualizer

### ISSUE THREAD

**Issue by SergioRAgostinho** [IntelVCL/Open3D#715]
Hey guys,

**Is your feature request related to a problem? Please describe.**
I'm making use of Open3D within Python, to save a couple of images from the visualizer upon pressing a key. Since I don't want to overwrite my image every time I press the "capture key", I need a variable visible inside my callback and incremented in every call, whose state persists between callback invocations. Check the following snippet
```python
def capture_image(vis):
    path = "temp_{:02d}.jpg".format(img_i)
    vis.capture_screen_image(path)
    print("Imaged saved to", path)
    img_i += 1
    return False
```

In regular Python, I believe that if the interpreter doesn't find a variable withing the local scope, it searches for it on its parent scope and upwards i.e.
```python
def what_is_the_value_of_img_i(some_other_argument):
    print(img_i)

img_i = 1
what_is_the_value_of_img_i(None) # prints 1
```
Since we are working with C/C++ wrapped in python, this trick doesn't really work. The function is not really aware of the "global" scope and you get a
```
UnboundLocalError: local variable 'img_i' referenced before assignment
```
 

**Describe the solution you'd like**
I'm not experience with Python bindings so I'm afraid I can't help much here. In C/C++ it is customary to provide a `void*` argument so that the user can pass additional data to the callback. I have no idea on what is the right way of porting such behavior to the Python layer.

**Describe alternatives you've considered**
I tried at some point to make use of the `key_to_callback_` attribute of 
https://github.com/IntelVCL/Open3D/blob/d993d21ff650227c736b08b0e7028f4c5ce16830/src/Visualization/Visualizer/VisualizerWithKeyCallback.h#L57
but it's unfortunately it is not exposed. The original idea was to make use of negative integer keys to store the data I needed to persist between invocations. Ultimately I noticed this would not also work because the map can only store a Callable type. :')

I also looked into the attributes of the (parent) Visualizer class but there's nothing there I can overwrite safely to write my data.

**Additional context**
n.a.

**Comment by dbadrian** [comment]
It can find the variable, but you need to declare it *nonlocal* inside the callback function (and probably define it before the callback.

```python
def rec_image(pcd):
    img_i = 0

    def save_image(vis):
        nonlocal img_i
        path = "temp_{:02d}.jpg".format(img_i)
        vis.capture_screen_image(path)
        img_i += 1

    vis = open3d.VisualizerWithKeyCallback()
    register_key_callback(vis, "v", save_image)

    vis.create_window("Record Images")
    vis.add_geometry(pcd)
    vis.run()  # until user presses "q" to terminate
    vis.destroy_window()
```

**Comment by SergioRAgostinho** [comment]
Indeed. Lack of experience on my part. It could be solved using either `nonlocal` or `global`.

Well, my feature request is still somewhat applicable to the C/C++ interface, so I'm unsure if I should close this issue or not.
